tools/src/main/scala/com/nvidia/spark/rapids/tool/planparser/SQLPlanParser.scala

```diff
@@ -0,0 +1,116 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.tool.planparser
+
+import scala.collection.mutable.ArrayBuffer
+
+import com.nvidia.spark.rapids.tool.qualification.PluginTypeChecker
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.execution.SparkPlanInfo
+import org.apache.spark.sql.execution.ui.{SparkPlanGraph, SparkPlanGraphCluster, SparkPlanGraphNode}
+import org.apache.spark.sql.rapids.tool.AppBase
+
+case class ExecInfo(
+    sqlID: Long,
+    exec: String,
+    expr: String,
+    speedupFactor: Int,
+    duration: Option[Long],
+    nodeId: Long,
+    isSupported: Boolean,
+    children: Option[Seq[ExecInfo]]) {
+  private def childrenToString = {
+    val str = children.map { c =>
+      c.map("       " + _.toString).mkString("\n")
+    }.getOrElse("")
+    if (str.nonEmpty) {
+      "\n" + str
+    } else {
+      str
+    }
+  }
+  override def toString: String = {
+    s"exec: $exec, expr: $expr, sqlID: $sqlID , speedupFactor: $speedupFactor, " +
+      s"duration: $duration, nodeId: $nodeId, " +
+      s"isSupported: $isSupported, children: " +
+      s"${childrenToString}"
+  }
+}
+
+case class PlanInfo(
+    sqlID: Long,
+    execInfo: Seq[ExecInfo]
+)
+
+object SQLPlanParser extends Logging {
+
+  def parseSQLPlan(
+      planInfo: SparkPlanInfo,
+      sqlID: Long,
+      checker: PluginTypeChecker,
+      app: AppBase): PlanInfo = {
+    val planGraph = SparkPlanGraph(planInfo)
+    // we want the sub-graph nodes to be inside of the wholeStageCodeGen so use nodes
+    // vs allNodes
+    val execInfos = planGraph.nodes.flatMap { node =>
+      parsePlanNode(node, sqlID, checker, app)
+    }
+    PlanInfo(sqlID, execInfos)
+  }
+
+  def parsePlanNode(
+      node: SparkPlanGraphNode,
+      sqlID: Long,
+      checker: PluginTypeChecker,
+      app: AppBase
+  ): Seq[ExecInfo] = {
+    node match {
+      case w if (w.name.contains("WholeStageCodegen")) =>
+        WholeStageExecParser(w.asInstanceOf[SparkPlanGraphCluster], checker, sqlID, app).parse
+      case f if (f.name == "Filter") =>
+        FilterExecParser(f, checker, sqlID).parse
+      case p if (p.name == "Project") =>
+        ProjectExecParser(p, checker, sqlID).parse
+      case o =>
+        logDebug(s"other graph node ${node.name} desc: ${node.desc} id: ${node.id}")
+        ArrayBuffer(ExecInfo(sqlID, o.name, expr = "", 1, duration = None, o.id,
+          isSupported = false, None))
+    }
+  }
+
+  // integer division if we want to roundup need to change
+  def average(arr: ArrayBuffer[Int]): Int = if (arr.isEmpty) 1 else arr.sum/arr.size
```

Comment:
if the sum of elements is 0 for an empty set, would avg(empty)=0 make more sense?

nit: spaces around `/`

Reply:
In  this case it's used with a speedup factor and applied to a duration , so your duration isn't going to go to 0, its going to stay the same so for empty array 1 makes sense, if its non-empty the sum should never be 0 as the speed-factor for individual things would be at least 1.

 Probably need a better name and documentation here so its clear and not used for other things